alexandria/core/views.py

```diff
@@ -168,87 +171,63 @@ def multi(self, request, **kwargs):
 
             return response
 
-    @action(methods=["post"], detail=True, url_path="generate-thumbnail")
-    def generate_thumbnail(self, request, pk=None):
-        if not settings.ALEXANDRIA_ENABLE_THUMBNAIL_GENERATION:
-            raise ValidationError(_("Thumbnail generation is not enabled."))
-
-        file = self.get_object()
-        if file.variant == models.File.THUMBNAIL:
-            raise ValidationError(
-                _("File is already a thumbnail, cannot generate thumbnail.")
-            )
-        elif file.renderings.all().exists():
-            raise ValidationError(
-                _("File already has thumbnail, cannot generate multiple thumbnails.")
+    def create(self, request, *args, **kwargs):
+        serializer = self.get_serializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        if settings.ALEXANDRIA_ENABLE_AT_REST_ENCRYPTION:
+            serializer.validated_data[
+                "encryption_status"
+            ] = settings.ALEXANDRIA_ENCRYPTION_METHOD
+        obj = serializer.save()
+        if (
+            settings.ALEXANDRIA_ENABLE_THUMBNAIL_GENERATION
+            and obj.variant == models.File.Variant.ORIGINAL
+        ):
+            try:
+                create_thumbnail(obj)
+            except DjangoCoreValidationError as e:
+                raise ValidationError(*e.messages)
+        if obj.variant == models.File.Variant.THUMBNAIL:
+            try:
+                # coerce the uploaded file to a thumbnail
+                create_thumbnail(obj)
+            except DjangoCoreValidationError as e:
+                # remove the uploaded file on failure to avoid
+                # collsions
+                obj.delete()
+                raise ValidationError(*e.messages)
+            # if the original already had a thumbnail remove that
+            self.queryset.filter(original=obj.original).exclude(pk=obj.pk).delete()
```

Comment:
This should probably be moved to the top of the `if` block, no? Otherwise it will delete the just-created thumbnail

Reply:
That's the indended behavior in my opinion. If it fails to thumbnail the thumbnail from the uploaded file we don't want it laying around. it could be anything, no?

Reply:
~~That's the indended behavior in my opinion. If it fails to thumbnail the thumbnail from the uploaded file we don't want it laying around. it could be anything, no?~~

sorry. I thought you were talking about L197. Since the whole `if` is suitable to replace a thumbnail I would not remove the old before the new one was successfully created. The `.exclude(pk=obj.pk)` should prevent deletion of the new thumbnail